packages/uma-auth-client/src/hooks/useOAuth.tsx

```diff
@@ -0,0 +1,100 @@
+import {auth, nwc, webln} from '@getalby/sdk';
+import { useEffect, useState } from "react";
+import { getUmaDomain } from "src/utils/getUmaDomain";
+import { useUser } from "./useUser";
+import { create } from 'zustand';
+import { persist } from "zustand/middleware";
+import { getNwcUmaDomain } from 'src/utils/getNwcUmaDomain';
+
+type UmaAuthToken = {
+  access_token: string,
+  refresh_token: string,
+  expires_at: number,
+  nwc_connection_uri: string;
+  commands: string[];
+  budget: string;
+  nwc_expires_at: string;
+};
+
+interface TokenState {
+  accessToken: string;
+  refreshToken: string;
+  expiresAt: number;
+}
+
+interface OAuthState {
+  codeVerifier?: string;
+  token?: TokenState;
+  identityNpub?: string;
+  identityRelayUrl?: string;
+  redirectUri?: string;
+  nwcConnectionUri?: string;
+  setToken: (token: TokenState) => void;
+  performOAuth: (uma: string) => Promise<void>;
+}
+
+export const useOAuth = create<OAuthState>()(
+  persist(
+    (set, get) => ({
+      setToken: (token) => set({
+        token,
+      }),
+      performOAuth: async (uma) => {
+        const state = get();
+        const result = await performOAuth(state, uma);
+        set(result);
+      },
+    }),
+    {
+      name: "uma-connect",
+      partialize: (state) => ({
+        nwcConnectionUri: state.nwcConnectionUri,
+      })
+    }
+  )
+);
+
+const performOAuth = async (state: OAuthState, uma: string) => {
+  const { identityNpub, identityRelayUrl, redirectUri, codeVerifier, token, nwcConnectionUri } = state;
+
+  // If we have a connection URI and a token that hasn't expired, we don't need to do anything
+  if (nwcConnectionUri && token && (!token.expiresAt || token.expiresAt < Date.now())) {
+    return state;
+  }
+
+  const authClient = new auth.OAuth2User({
+    client_id: `${identityNpub} ${identityRelayUrl}`,
+    callback: redirectUri!, // TODO: get this from client app
+    scopes: [],
+    user_agent: "uma-connect",
+    request_options: {
+      base_url: getNwcUmaDomain(uma),
+    },
+  });
+
+  let resultToken: UmaAuthToken;
+  if (token?.refreshToken) {
+    authClient.token = {
+      access_token: token.accessToken,
+      refresh_token: token.refreshToken,
+      expires_at: token.expiresAt,
+    };
+
+    resultToken = (await authClient.refreshAccessToken()).token as UmaAuthToken;
+  } else {
+    authClient.code_verifier = codeVerifier!; // TODO: where do we get this?
+    const code = new URLSearchParams(redirectUri!).get("code") || "";
```

Comment:
This isn't the right place to pull the code. You get the code back from the oauth service when it redirects back to redirectUri. Since in this function, you're doing the token exchange, that should mean that it was already passed back here as a url param in window.location. [For reference](https://github.com/jklein24/bitcoin-connect/blob/c3de26d3ae2799d5b5772811b4b39f3889009319/src/connectors/UMANWCConnector.ts#L32)

Reply:
If the client app doesn't maintain the code query param in the url then won't this be lost? we'd have to get it from the client app providing the original uri that the oauth flow redirected to (which should include the code and state?)

Reply:
Yep... this is the challenge with oauth being redirect-based. You need to get the code from the URL where the redirect goes.